API REMOTE_USER authentication

dLobatog opened the pull request:
This pull request allows the setting 'authorize_login_delegation_api' to work. Basically if foreman is behind any kind of proxy using Integrated Windows Authentication or Apache and mod_authnz_ldap, CAS, Cosign, WebAuth, mod_auth_sspi, etc.., foreman should trust its authentication through REMOTE_USER on API calls too.

While working on this I noticed there is a major overlap between this file and application_controller, possible next steps include moving Oauth to app/services/sso, and factoring out the authorization/authentication code from application_controller and services/api/authorization.rb into a single authentication/authorization file.

dLobatog commented:
[test]

domcleal commented:
@ares could you review please?  My initial reaction is duplication of SSO::Apache is bad...

ares commented:
Agreed. @eLobato Could you please reuse app/services/sso/apache.rb? If you don't want to cleanup clash between ApplicationController and BaseController you can think of ApplicationController#require_login as a matching for BaseController#authorize. So I'd use the same workflow for SSO. Could you define a module which would be included in both and would:
- find available SSO (see ApplicationController#available_sso
- try to authenticated using found SSO method
- use it in BaseController as an alternative for Api::Authorization#authenticate (oauth and basic_auth would be used only if SSO fails)
- modify ApplicationController to use a new module

dLobatog committed:
API and Regular authentication refactor
Regular authentication had SSO, and http basic auth. API had Oauth
and http basic. This basically merges them both into a single file,
since there was a lot of overlap between application_controller and
api base_controller. A major benefit this brings is the possiblity of
using Apache REMOTE_USER authentication through the API which was
not possible before.

Assert_response is the proper way instead of assert_equal

Moved Basic and Oauth to SSO

Logout url in Apache

Code cleanups

Basic Authentication tests

OAuth tests

Add login_url to be the default url services redirect to

Fix OAuth test 'authenticate is user.current is not set'

Make api and base controller use 'login_url' from SSO service if they need to redirect to them

Signo test tests controller redirect not SSO redirect

Only let apache work if REMOTE_USER is set. Otherwise it does not make sense to let it try to authenticate without even an username

Return username on authentication errors

Changed apache_logout_url by something more generic, login_delegation_logout_url

dLobatog commented:
@ares I believe I did what you wanted, basically refactored them both into a single module which is used by application_controller and api/base_controller. Main benefit we gain for this (aside from clarity) is the ability of using SSO for api requests.

dLobatog commented:
@ares I just added logout_url to Apache and moved Oauth and HttpBasic to a separate module in /services/sso

ares commented:
@eLobato in general it looks good to me, I made a PR with few small changes to your branch, please squash them to your PR. I also tested all supported methods and it seems to work smoothly. Could you also please add tests for basic and oauth authentication methods?

Thank you for great amount of work you have done on this so far!

dLobatog commented:
@ares  There you go!

domcleal commented on app/services/sso/basic.rb:
```diff
@@ -0,0 +1,21 @@
+
+    def authenticate!
+      user = controller.authenticate_with_http_basic { |u, p| User.try_to_login(u, p) }
+      self.user = user.login if user.present?
```
Should this just be `user`, not `user.login`?

ares replied:
`user.login` is correct, user attribute should hold just login not user object
however to make an authentication we must find the user instance on previous line

domcleal replied:
Ah, you're right.  I thought I saw an instance where it was an object, but I think I'm mistaken - I can't find it now.

domcleal commented on app/services/sso/oauth.rb:
```diff
@@ -0,0 +1,40 @@
+module SSO
+  class Oauth < Base
+    def available?
+      !!(request.authorization =~ /^OAuth/)
```
Maybe `controller.api_request? &&` too?

domcleal commented on app/services/sso/oauth.rb:
```diff
@@ -0,0 +1,40 @@
+      end
+
+      unless (incoming_key = OAuth::RequestProxy.proxy(request).oauth_consumer_key) ==
+        Setting['oauth_consumer_key']
```
Nitpick, odd line wrapping here, probably best to join it to the previous line or indent

domcleal commented on app/services/sso/oauth.rb:
```diff
@@ -0,0 +1,40 @@
+    end
+
+    def authenticated?
+      self.user = (User.current.present? ? User.current : authenticate!).login
```
If `authenticate!` returns nil (failure), this'll try to call `nil.login` and fail

domcleal commented on app/controllers/concerns/foreman/controller/authentication.rb:
```diff
@@ -0,0 +1,74 @@
+        return !User.current.nil?
+      else
+        if api_request?
+          # Deprecation warning: This is only triggered on ?format.json ?format.yaml calls (deprecated 'API')
```
Is that right, isn't it on all API requests?

dLobatog replied:
That would trigger a double render error since 'new' API requests redirects are handled on its base controller https://github.com/theforeman/foreman/blob/develop/app/controllers/api/base_controller.rb

domcleal replied:
Right.. I don't see how this is restricted to the old API though.  Both APIs return true for `api_request?` I think.

domcleal commented on app/controllers/concerns/foreman/controller/authentication.rb:
```diff
@@ -0,0 +1,74 @@
+        else
+          # Keep the old request uri that we can redirect later on
+          session[:original_uri] = request.fullpath
+          (redirect_to login_users_path and return) unless Setting['signo_sso']
```
Why `unless Setting['signo_sso']`?

dLobatog replied:
Because if Setting['signo_sso'] is set, we should redirect to SSO::Signo.login_url

domcleal replied:
Hm, the SSO abstraction is a bit broken here then, isn't it... can we make `.login_url` a method on all SSO objects and always do the render in this mixin?  Then it'll be consistent and we won't get renders from within the SSO object.

domcleal replied:
(the default .login_url would simply be the login_users_path)

domcleal commented on app/controllers/concerns/foreman/controller/authentication.rb:
```diff
@@ -0,0 +1,74 @@
+    end
+  end
+
+  def authorized
```
Maybe `authorized?` for consistency

domcleal commented:
Thanks for the big refactor @eLobato and @ares, I think this is much more sustainable.  My main concern is lack of tests for the new SSO classes, could you look at adding some?  There should also be some functional tests for the API REMOTE_USER authentication itself, similar to the ones in test/functional/hosts_controller_test.rb.

What do you think @isratrade?

dLobatog commented:
@domcleal I just added a few more tests so that we have tests for all 'SSO' classes.

domcleal commented on test/unit/sso/oauth_test.rb:
```diff
@@ -0,0 +1,40 @@
+    User.current = nil
+    oauth = SSO::Oauth.new(get_controller(true))
+    oauth.expects(:authenticate!).times(1)
+    oauth.authenticated?
```
missing assertion for the return value?

dLobatog commented:
I just fixed a few minor things like getting rid of the 'render unauthorized' for the old "API"

dLobatog commented:
Can you check the build? It seems extremely weird that it just breaks on Postgres 1.9.3

domcleal commented:
Ignore the build failure, it was an issue elsewhere.

domcleal commented:
I've been testing this before merging and hit a couple of issues.  Would you mind fixing them and adding tests where possible?

1) Using an incorrect password under plain auth with the new API fails with an error:

<pre>
$ curl -u admin:incorrect -H "Accept: application/json" http://0.0.0.0:3000/api/hosts
{"error":{"message":"undefined local variable or method `user_login' for #<Api::V1::HostsController:0x000000064c4ae0>"}}
logs:
invalid user
Setting current user thread-local variable to nil
SSO failed, falling back to login form
undefined local variable or method `user_login' for #<Api::V1::HostsController:0x000000064c4ae0> (NameError)
/home/dcleal/code/foreman/foreman/app/controllers/api/base_controller.rb:77:in `authorize'
</pre>

2) Using an incorrect user under plain auth with either the new or old API, or on the usual login page appears to hang.  No HTTP response, not sure what the server's doing at this point.

I'm also getting the hangs on a Passenger setup while trying to test the API with REMOTE_USER, but I guess that's related to (2).

domcleal commented on app/models/setting/auth.rb:
```diff
@@ -18,7 +18,8 @@ def self.load_defaults
         self.set('signo_sso', N_('Use Signo SSO for login'), false),
-        self.set('signo_url', N_('Signo SSO url'), "https://#{fqdn}/signo")
+        self.set('signo_url', N_('Signo SSO url'), "https://#{fqdn}/signo"),
+        self.set('apache_logout_url', N_('Redirect your users to this url on logout (Apache SSO should also be enabled)'), 'https://logouturl')
```
@ohadlevy mentioned on PR #840 that this could have a more generic name, perhaps `authorize_login_delegation_logout_url`?

dLobatog replied:
Agreed. I will just call it logout_url and make a more generic explanation, since this is not restricted to Apache.

domcleal commented:
Thanks for this great refactor @eLobato, merged as 852ff71 for Foreman 1.3.

I discovered what was causing the hangs - an LDAP auth source pointing at a non-existent host!
